### PR TITLE
Maintain Chapter 6-2

### LINKED ISSUE #563: Maintain Chapter 6-2 The match control flow construct
- [x] `match` formatting not consistent in the introduction
- [x] remove use core::fmt;
- [x] Coin::Penny(_)  should be Coin::Penny in text
- [x] complete with matching integers / non exhaustive arms 2.5.0
- [x] arms don't need to be in the same order anymore
- [x] `or` expressions between variants
- [x] matching over tuple of enums
- [x] remove useless --available-gas=200000000 flag

### PR DESCRIPTION
**Pull request by TAdev0**
Resolves #563

- - -
This change is [Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/575)

### CODE REVIEW

File: listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo
Comment (enitrat) [inline listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo:1]: Same, all examples of implementations of the `Print` trait must be removed and replaced with something else

File: src/ch06-02-the-match-control-flow-construct.md [hunk 1]
```diff
@@ -1,34 +1,34 @@
...
+Let’s break down the `match` expression in the `value_in_cents` function. First we list the `match` keyword followed by an expression, which in this case is the value `coin`. This seems very similar to a conditional expression used with if, but there’s a big difference: with if, the condition needs to evaluate to a Boolean value, but here it can be any type. The type of `coin` in this example is the `Coin` enum that we defined on the first line.
 
-Next are the `match` arms. An arm has two parts: a pattern and some code. The first arm here has a pattern that is the value `Coin::Penny(_)` and then the `=>` operator that separates the pattern and the code to run. The code in this case is just the value `1`. Each arm is separated from the next with a comma.
+Next are the `match` arms. An arm has two parts: a pattern and some code. The first arm here has a pattern that is the value `Coin::Penny` and then the `=>` operator that separates the pattern and the code to run. The code in this case is just the value `1`. Each arm is separated from the next with a comma. The order of arms and their patterns is not enforced to match the order in the enum definition.
```
Comment (enitrat) [inline src/ch06-02-the-match-control-flow-construct.md:17]:
```suggestion
Next are the `match` arms. An arm has two parts: a pattern and some code. The first arm here has a pattern that is the value `Coin::Penny` and then the `=>` operator that separates the pattern and the code to run. The code in this case is just the value `1`. Each arm is separated from the next with a comma.
```
useless imo, + in some cases you actually have to

**Review by TAdev0 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/575)* status: all files reviewed, 7 unresolved discussions (waiting on @enitrat)

___
*[`listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo` line 0 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/575#gh-1477821116:-NpzroveEqdZJkxbGPkh:b-d02py9) ([raw file](https://github.com/cairo-book/cairo-book/blob/4cd73cd746013ac0e859628b2fb18e61d2f547fc/listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo#L0)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

Same, all examples of implementations of the `Print` trait must be removed and replaced with something else
</blockquote></details>

yes, there was just one call to ''print'' in  `state.print();` 
I created nested match to print the states , i think its nice to keep this example like this
___
*[`listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo` line 5 at r3](https://reviewable.io/reviews/cairo-book/cairo-book/575#-Npyt86D-R4RoS5OfjaR:-Npzs4R31e0EADpuF0e2:b-rvxlwo) ([raw file](https://github.com/cairo-book/cairo-book/blob/f392079cfac7cbaf573a6fb08a7711f7ed3478ce/listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo#L5)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

what for?
</blockquote></details>

a mistake, removed
___
*[`listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo` line 30 at r3](https://reviewable.io/reviews/cairo-book/cairo-book/575#-NpytZhTCwleYQ9CTrR_:-NpzsVtxCjRscnNHK47O:b-g6r69s) ([raw file](https://github.com/cairo-book/cairo-book/blob/f392079cfac7cbaf573a6fb08a7711f7ed3478ce/listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo#L30)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

Coin::Quarter(state) => { println!("State quarter from {:?}!", state); 25 }
</blockquote></details>

should we keep the nested match pattern with harcoded printed value, or introduce debug to print state?
It seems enum doesnt have implementation of Display (this is the error i get)
___
*[`src/ch06-02-the-match-control-flow-construct.md` line 17 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/575#gh-1477822698:-NpzwsreBe59guTQYW-1:b-bb3k2w) ([raw file](https://github.com/cairo-book/cairo-book/blob/4cd73cd746013ac0e859628b2fb18e61d2f547fc/src/ch06-02-the-match-control-flow-construct.md#L17)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

```suggestion
Next are the `match` arms. An arm has two parts: a pattern and some code. The first arm here has a pattern that is the value `Coin::Penny` and then the `=>` operator that separates the pattern and the code to run. The code in this case is just the value `1`. Each arm is separated from the next with a comma.
```
useless imo, + in some cases you actually have to
</blockquote></details>

already updated in last commit
___
*[`src/ch06-02-the-match-control-flow-construct.md` line 49 at r3](https://reviewable.io/reviews/cairo-book/cairo-book/575#-NpytPfW3o-lJgXLqFvO:-NpzxLW03mkb-cCMODGK:baoilr7) ([raw file](https://github.com/cairo-book/cairo-book/blob/f392079cfac7cbaf573a6fb08a7711f7ed3478ce/src/ch06-02-the-match-control-flow-construct.md#L49)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

We can then use that binding in the `println!` expression, thus getting the inner state value out of the `Coin` enum variant for `Quarter`.
</blockquote></details>

tell me which pattern we should choose
___
*[`src/ch06-02-the-match-control-flow-construct.md` line 53 at r3](https://reviewable.io/reviews/cairo-book/cairo-book/575#-NpytfPqC_3-mnjWt1Rx:-NpzxYKCDoQL3fz4CbYe:b2ni9ah) ([raw file](https://github.com/cairo-book/cairo-book/blob/f392079cfac7cbaf573a6fb08a7711f7ed3478ce/src/ch06-02-the-match-control-flow-construct.md#L53)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

not required imo
</blockquote></details>

ok
___
*[`src/ch06-02-the-match-control-flow-construct.md` line 85 at r3](https://reviewable.io/reviews/cairo-book/cairo-book/575#-Npyu0Ce8QPt2yjdF0jy:-Npzy0VQ1ncWMjwb3g2q:bfbuhl6) ([raw file](https://github.com/cairo-book/cairo-book/blob/f392079cfac7cbaf573a6fb08a7711f7ed3478ce/src/ch06-02-the-match-control-flow-construct.md#L85)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

I think it's bad that the program fails. Can you replace it with the one from the rust book? 

```rust
fn main() {
    fn plus_one(x: Option<i32>) -> Option<i32> {
        match x {
            None => None,
            Some(i) => Some(i + 1),
        }
    }

    let five = Some(5);
    let six = plus_one(five);
    let none = plus_one(None);
}
```
</blockquote></details>

done. I was a bit wrong , the programs compiles correctly, but fails in its execution .
___
*[`src/ch06-02-the-match-control-flow-construct.md` line 162 at r3](https://reviewable.io/reviews/cairo-book/cairo-book/575#-NpyuJ7U7Ix4He_6tcEK:-NpzypG04CIpR-N5S2Qe:brxtgn9) ([raw file](https://github.com/cairo-book/cairo-book/blob/f392079cfac7cbaf573a6fb08a7711f7ed3478ce/src/ch06-02-the-match-control-flow-construct.md#L162)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

there's one single felt size
</blockquote></details>

yeah sure, integers is plurial as well but we match on one single integer u128 max. I can rewrite

**Review by enitrat — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/575)* status: all files reviewed, 3 unresolved discussions (waiting on @TAdev0)

___
*[`listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo` line 30 at r3](https://reviewable.io/reviews/cairo-book/cairo-book/575#-NpytZhTCwleYQ9CTrR_:-Nq1XkpuCBpF0U6r0K7J:b-j077mb) ([raw file](https://github.com/cairo-book/cairo-book/blob/f392079cfac7cbaf573a6fb08a7711f7ed3478ce/listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo#L30)):*
<details><summary><i>Previously, TAdev0 (Tristan) wrote…</i></summary><blockquote>

should we keep the nested match pattern with harcoded printed value, or introduce debug to print state?
It seems enum doesnt have implementation of Display (this is the error i get)
</blockquote></details>

We should derive debug on the Enum. It should work fine with debug, as the debug trait should display both the enum name and the enum variant used (the purpose of it is to display stuff in an "ugly" debug way)
___
*[`listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo` line 3 at r4](https://reviewable.io/reviews/cairo-book/cairo-book/575#-Nq1YXgr1BmEfS9w5VUt:-Nq1YXgr1BmEfS9w5VUu:undefined) ([raw file](https://github.com/cairo-book/cairo-book/blob/55dfd43d475cc5839b309846acd5ded544227da4/listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo#L3)):*
> ```Cairo
> // ANCHOR: enum_def
> 
> #[derive(Drop)]
> ```





_Suggestion:_
```Cairo
#[derive(Copy, Drop, Debug)] // Debug so we can inspect the state in a minute
enum UsState {
```

**Review by TAdev0 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/575)* status: 1 of 3 files reviewed, 3 unresolved discussions (waiting on @enitrat)

___
*[`listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo` line 30 at r3](https://reviewable.io/reviews/cairo-book/cairo-book/575#-NpytZhTCwleYQ9CTrR_:-Nq2R43SBfL9J2UrJQGB:byhzbqh) ([raw file](https://github.com/cairo-book/cairo-book/blob/f392079cfac7cbaf573a6fb08a7711f7ed3478ce/listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo#L30)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

We should derive debug on the Enum. It should work fine with debug, as the debug trait should display both the enum name and the enum variant used (the purpose of it is to display stuff in an "ugly" debug way)
</blockquote></details>

ok! 
I ran this example in a main function and the output looks like this : 
```
State quarter from UsState::Alabama(())!
Run completed successfully, returning [25]
```
Indeed its a bit ugly :p
___
*[`listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo` line 3 at r4](https://reviewable.io/reviews/cairo-book/cairo-book/575#-Nq1YXgr1BmEfS9w5VUt:-Nq2QkOZ-cDN1bL4bxCI:byi5zzt) ([raw file](https://github.com/cairo-book/cairo-book/blob/55dfd43d475cc5839b309846acd5ded544227da4/listings/ch06-enums-and-pattern-matching/no_listing_06_match_pattern_bind/src/lib.cairo#L3)):*
> ```Cairo
> // ANCHOR: enum_def
> 
> #[derive(Drop)]
> ```

We don't need Copy here, why did you add it?

**Review by enitrat — APPROVED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/575)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @TAdev0)

**Pull request merged**